File: Golang/reverse_LinkedList.go
package main
import "fmt"
type Node struct {
   value int
   next *Node
}
func NewNode(value int, next *Node) *Node{
   var n Node
   n.value = value
   n.next = next
   return &n
}
func TraverseLinkedList(head *Node){
   fmt.Printf("Input Linked List is: ")
   temp := head
   for temp != nil {
      fmt.Printf("%d ", temp.value)
      temp = temp.next
   }
   fmt.Println()
}
func ReverseLinkedList(head *Node){
   if head == nil{
      return
   }
   ReverseLinkedList(head.next)
   fmt.Printf("%d ", head.value)
}
func main(){
   head := NewNode(30, NewNode(10, NewNode(40, NewNode(40, nil))))
   TraverseLinkedList(head)
   fmt.Printf("Reversal of the input linked list is: ")
   ReverseLinkedList(head)
}
Output

File: CPP/reverse_linked_list.cpp
#include <iostream>

using namespace std;

class SinglyLinkedListNode
{
public:
  int data;
  SinglyLinkedListNode *next;

  SinglyLinkedListNode(int node_data)
  {
    this->data = node_data;
    this->next = nullptr;
  }
};

class SinglyLinkedList
{
public:
  SinglyLinkedListNode *head;
  SinglyLinkedListNode *tail;

  SinglyLinkedList() {
    this->head = nullptr;
    this->tail = nullptr;
  }

  void insert_node(int node_data) {
    SinglyLinkedListNode *node = new SinglyLinkedListNode(node_data);

    if (!this->head) {
      this->head = node;
    } else {
      this->tail->next = node;
    }

    this->tail = node;
  }
};

SinglyLinkedListNode *reverse(SinglyLinkedListNode *llist, SinglyLinkedListNode *prev = nullptr)
{
  SinglyLinkedListNode *next = llist->next;
  llist->next = prev;
  if (!next) return llist;
  return reverse(next, llist);
}

int main() {
  int T;
  cin >> T;
  while (T--) {
    SinglyLinkedList *llist = new SinglyLinkedList();

    int llist_count;
    cin >> llist_count;
    for (int i = 0; i < llist_count; i++) {
      int llist_item;
      cin >> llist_item;
      llist->insert_node(llist_item);
    }
    SinglyLinkedListNode *llist1 = reverse(llist->head);
    SinglyLinkedListNode *llist1Free = llist1;
    while (llist1) {
      cout << llist1->data;
      llist1 = llist1->next;
      if (llist1) {
        cout << " ";
      }
    }
    cout << "\n";

    while (llist1Free) {
      SinglyLinkedListNode *temp = llist1Free;
      llist1Free = llist1Free->next;
      free(temp);
    }
  }

  return 0;
}
